src/v/utils/token_bucket.h

```diff
@@ -79,9 +80,23 @@ class token_bucket {
         }
 
         if (_rate > new_rate) {
-            _sem.consume(_rate - new_rate);
+            if (_sem.current() <= _rate) {
+                _sem.consume(_rate - new_rate);
+            } else {
+                /*
+                 * if current > rate it means that we have accumulated burst
+                 * we should change current to new_rate to start accumulation
+                 * from begining
+                 */
+                _sem.consume(_sem.current() - new_rate);
+            }
         } else {
-            _sem.signal(new_rate - _rate);
+            if (_sem.current() <= _rate) {
+                _sem.signal(new_rate - _rate);
+            } else {
+                // Same as previous comment
+                _sem.consume(_sem.current() - new_rate);
```

Comment:
is this supposed to be `signal`?

Reply:
No, we make token amount equal to new_rate.
If there now sem.current() and sem.current() > new_rate, we need to consume _sem.current() - new_rate